pylint/checkers/base/function_checker.py

```diff
@@ -0,0 +1,222 @@
+        self, node: nodes.FunctionDef
+    ) -> None:
+        """Check a FunctionDef to find context-managers and generators used in a no
+        cleanup pattern.
+
+        :param node: FunctionDef node to check
+        :type node: nodes.FunctionDef
+        """
+        # if function does not use a Yield statement, it cant be a generator or a contextmanager
+        # TODO: should this also capture YieldFrom?
+        yield_nodes = list(node.nodes_of_class(nodes.Yield))
+        if not yield_nodes:
+            return
+
+        # check if function has a contextmanager decorator
+        contextmanager_str = "contextlib.contextmanager"
+        should_be_added_to_contextmanagers = any(
+            decorator
+            # this is fully qualified import name
+            for decorator in node.decoratornames()
+            if decorator == contextmanager_str
+        )
+        # cant be a contextmanager and a erroring generator at the same time
+        if (
+            should_be_added_to_contextmanagers
+            and self._node_fails_contextmanager_cleanup(node, yield_nodes=yield_nodes)
+        ):
+            self._add_node_to_contextmanagers(node)
+            return
+
+        # otherwise could be a generator that uses a contextmanager
+        with_nodes = list(node.nodes_of_class(nodes.With))
+        async_with_nodes = list(node.nodes_of_class(nodes.AsyncWith))
+        if not with_nodes and not async_with_nodes:
+            # no inner contextmanager, so no need to check
+            return
+        all_nodes = chain(with_nodes, async_with_nodes)
+        for with_node in all_nodes:
+            if self._node_fails_generator_checks(with_node):
+                self._add_node_to_generators(with_node, node)
+
+    @staticmethod
+    def _node_fails_contextmanager_cleanup(
+        node: nodes.FunctionDef, yield_nodes: list[nodes.Yield]
+    ) -> bool:
+        """Check if a node fails contextmanager cleanup.
+
+        Current checks for a contextmanager:
+            - only if the context manager yields something other than None
+            - only if the context manager lacks a finally, or does not catch GeneratorExit
+
+        :param node: Node to check
+        :type node: nodes.FunctionDef
+        :return: True if fails, False otherwise
+        :param yield_nodes: List of Yield nodes in the function body
+        :type yield_nodes: list[nodes.Yield]
+        :rtype: bool
+        """
+        # if context manager yields a non-constant value, then continue checking
+        if any(
+            yield_node.value is None or isinstance(yield_node.value, nodes.Const)
+            for yield_node in yield_nodes
+        ):
+            return False
+        # if function body has multiple Try, filter down to the ones that have a yield node
+        try_with_yield_nodes = [
+            try_node
+            for try_node in node.nodes_of_class(nodes.Try)
+            if list(try_node.nodes_of_class(nodes.Yield))
```

Comment:
nit: probably don't need to list-ify this generator

Reply:
Maybe I'm missing something obvious, but it seemed on a first impression that `list` is more expensive than `any` in this instance. Or is that not so?